File: 3-can-sum.py
# Solution3: Tabulation
from typing import List

# Solution3: Tabulation
# m: targetSum
# n: length of numbers list
# time: O(m * n)
# space: O(m)


def canSum(targetSum: int, numbers: List[int]) -> bool:
    table = [False for _ in range(targetSum+1)]
    table[0] = True

    for i in range(targetSum+1):
        if table[i]:
            for n in numbers:
                if i+n <= targetSum:
                    table[i+n] = True
    return table[targetSum]


print(canSum(7, [2, 3]))  # True
print(canSum(7, [5, 3, 4, 7]))  # True
print(canSum(7, [2, 4]))  # False
print(canSum(8, [2, 3, 5]))  # True
print(canSum(200, [7, 14]))  # False

# Solution2: Memoization
# def canSum(targetSum: int, numbers: List[int], memo: dict) -> bool:
#     if targetSum in memo:
#         return memo[targetSum]
#     if targetSum == 0:
#         return True
#     if targetSum < 0:
#         return False

#     for num in numbers:
#         remainder = targetSum - num
#         if canSum(remainder, numbers, memo):
#             memo[targetSum] = True
#             return True
#     memo[targetSum] = False
#     return False


# region Solution1: Brute Force
#
# def canSum(targetSum: int, numbers: List[int]) -> bool:
#     if targetSum == 0:
#         return True
#     if targetSum < 0:
#         return False
#     for num in numbers:
#         remainder = targetSum - num
#         if canSum(remainder, numbers):
#             return True
#     return False
# endregion

# print(canSum(7, [2, 3], {}))  # True
# print(canSum(7, [5, 3, 4, 7], {}))  # True
# print(canSum(7, [2, 4], {}))  # False
# print(canSum(8, [2, 3, 5], {}))  # True
# print(canSum(200, [7, 14], {}))  # False

File: 3-can-sum.js
// Solution3: Tabulation
// m: targetSum
// n: length of numbers array
// time: O(m * n)
// space: O(m)
const canSum = (targetSum, numbers) => {
  const table = Array(targetSum + 1).fill(false);

  table[0] = true;

  for (let i = 0; i <= targetSum; i++) {
    if (table[i] === true) {
      for (let n of numbers) {
        table[i + n] = true;
      }
    }
  }
  return table[targetSum];
};

// Solution2: Recursive + Memoization

// m: targetSum
// n: numbers length
// time: O(n * m), memo has m nodes, and each node will have n possible branches. so the time complexity will be O(m * n)
// space: O(m) - height of the tree is m. Normally, the recursive code complexity is the height of the tree

// const canSum = (targetSum, numbers, memo = {}) => {
//   if (targetSum in memo) return memo[targetSum];
//   if (targetSum === 0) return true;
//   if (targetSum < 0) return false;

//   for (let num of numbers) {
//     const remainder = targetSum - num;
//     if (canSum(remainder, numbers, memo) === true) {
//       memo[targetSum] = true;
//       return true;
//     }
//   }
//   memo[targetSum] = false;
//   return false;
// };

// Solution1: Recursive
// m: targetSum
// n: numbers length
// time: O(n^m)
// space: O(m)

// const canSum = (targetSum, numbers) => {
//   if (targetSum === 0) return true;
//   if (targetSum < 0) return false;

//   for (let num of numbers) {
//     const remainder = targetSum - num;
//     if (canSum(remainder, numbers) === true) {
//       return true;
//     }
//   }
//   return false;
// };

console.log(canSum(7, [2, 3])); // true
console.log(canSum(7, [5, 3, 4, 7])); // true
console.log(canSum(7, [2, 4])); // false
console.log(canSum(8, [2, 3, 5])); // true
console.log(canSum(300, [7, 14])); // false
